res/controllers/Numark-Mixtrack-Pro-FX-scripts.js

```diff
@@ -0,0 +1,746 @@
+            this.inKey = "pitch_down";
+        },
+        unshift: function() {
+            this.group = "[EffectRack1_EffectUnit" + deckNumber + "_Effect1]";
+            this.inKey = "next_effect";
+        }
+    });
+};
+
+MixtrackProFX.EffectUnit.prototype = new components.ComponentContainer();
+
+MixtrackProFX.Deck = function(number) {
+    components.Deck.call(this, number);
+
+    var channel = number - 1;
+
+    this.playButton = new components.PlayButton({
+        midi: [0x90 + channel, 0x00],
+        shiftControl: true,
+        sendShifted: true,
+        shiftOffset: 0x04
+    });
+
+    this.cueButton = new components.CueButton({
+        midi: [0x90 + channel, 0x01],
+        shiftControl: true,
+        sendShifted: true,
+        shiftOffset: 0x04
+    });
+
+    this.syncButton = new components.SyncButton({
+        midi: [0x90 + channel, 0x02],
+        shiftControl: true,
+        sendShifted: true,
+        shiftOffset: 0x01
+    });
+
+    this.pflButton = new components.Button({
+        type: components.Button.prototype.types.toggle,
+        midi: [0x90 + channel, 0x1B],
+        key: "pfl"
+    });
+
+    this.loadButton = new components.Button({
+        inKey: "LoadSelectedTrack"
+    });
+
+    this.volume = new components.Pot({
+        group: this.currentDeck,
+        inKey: "volume"
+    });
+
+    this.treble = new components.Pot({
+        group: "[EqualizerRack1_" + this.currentDeck + "_Effect1]",
+        inKey: "parameter3"
+    });
+
+    this.mid = new components.Pot({
+        group: "[EqualizerRack1_" + this.currentDeck + "_Effect1]",
+        inKey: "parameter2"
+    });
+
+    this.bass = new components.Pot({
+        group: "[EqualizerRack1_" + this.currentDeck + "_Effect1]",
+        inKey: "parameter1"
+    });
+
+    this.filter = new components.Pot({
+        group: "[QuickEffectRack1_" + this.currentDeck + "]",
+        inKey: "super1"
+    });
+
+    this.gain = new components.Pot({
+        inKey: "pregain"
+    });
+
+    this.pitch = new components.Pot({
+        inKey: "rate",
+        invert: true
+    });
+
+    this.padSection = new MixtrackProFX.PadSection(number);
+
+    this.shiftButton = new components.Button({
+        input: function(channel, control, value) {
+            // each shift button shifts both decks
+            // more consistent with the logic burned into hardware
+            if (value === 0x7F) {
+                MixtrackProFX.shifted = true;
+                MixtrackProFX.deck.shift();
+                MixtrackProFX.browse.shift();
+                MixtrackProFX.effect.shift();
+            } else if (value === 0x00) {
+                MixtrackProFX.shifted = false;
+                MixtrackProFX.deck.unshift();
+                MixtrackProFX.browse.unshift();
+                MixtrackProFX.effect.unshift();
+            }
+        }
+    });
+
+    this.loop = new components.Button({
+        outKey: "loop_enabled",
+        midi: [0x94 + channel, 0x40],
```

Comment:
Yes, I found it in some other mapping (it's not documented on https://github.com/mixxxdj/mixxx/wiki/Components-JS), however the button acts differently in Serato and VDJ when it comes to enabling the loop, and I tried to recreate that behavior.

`LoopToggleButton` toggles the loop without moving it. If current track position is after the loop OUT point, playhead is moved to the loop IN point. As far as I can tell, the behavior is the same when shifted (again, I can't find any documentation on how it's *supposed* to work). In this case user can only set a new loop position with the loop IN/OUT buttons, so can't easily create a loop with a "quantized" size (1/2/4/etc beats) (correct me if I'm wrong).

In Serato/VDJ the button enables the loop, setting the IN point at current playback position. Shift+loop resets playback position to loop IN point. I think "reloop" is [treated differently in Mixxx](https://manual.mixxx.org/2.3/en/chapters/appendix/mixxx_controls.html#control-[ChannelN]-reloop_toggle) though, i.e. it behaves the same as `LoopToggleButton`, so that's what shift+loop does in my mapping.

Reply:
Ok, the COs being used be the LoopToggleButton are also deprecated so I guess its not a good idea to use them. 

From the Mixtrack manual:
> Loop  On/Off:  Press  to  activate  auto  loop  on/off.  Hold  Shift  and  press this  button  to  trigger  a  reloop.

from what I can tell, you'll want the following configuration:
```javascript
outKey: "loop_enabled",
unshift: function() {
    this.inKey = "beatloop_activate",
},
shift: function() {
    this.inKey = "reloop_toggle",
}
```

Reply:
If I understood you correctly, you may want to use [`loop_in_goto`](https://manual.mixxx.org/2.3/en/chapters/appendix/mixxx_controls.html#control-[ChannelN]-loop_in_goto) instead of reloop toggle. No loop toggling and independent of play position.

Reply:
The problem with `beatloop_activate` is that it only activates the loop - pressing it again does not disable the loop. That's why in current code I trigger `beatlooproll_activate` to disable the loop (when loop is enabled).

To me there are two possible solutions here:
1. What @Swiftb0y proposes, i.e. `beatloop_activate` and `reloop_toggle` when shifted. This way loop IN can be placed on current playback position, and loop can be disabled with shift+loop.
2. Current code and `loop_in_goto` when shifted. This is the same behavior as in Serato/VDJ.

Reply:
I don't care either way. Mixxx does not recommend a standard behavior (like we do for the play and sync buttons for instance), so implement whatever you think is better as long as it confirms to the hardware vendors manual (and then document the specific behavior in our own manual page).

Reply:
Ok, in that case I'm leaving current code.

Reply:
A user [complained about this on the forum](https://mixxx.discourse.group/t/numark-mixtrack-pro-fx/19561/57) and I agree this design is odd. I'm not concerned with following how it works in Serato exactly; let's focus on what's useful. I suggest that when this button is shifted, check if `loop_enabled` is true. If `loop_enabled` is false, then use `reloop_toggle`. If `loop_enabled` is true, use `reloop_andstop`.

Reply:
Isn't it a it unintuitive that a "reloop" causes the playback to stop (`reloop_andstop`)?

What do you thing about this approach: 
```
if (shifted) {
    script.triggerControl(group, "loop_in_goto");

    if (engine.getValue(group, "loop_enabled") === 0) {
        script.triggerControl(group, "reloop_toggle");
    }
}
```

Reply:
> Isn't it a it unintuitive that a "reloop" causes the playback to stop (reloop_andstop)?

No, it's quite intuitive and useful. Try it. It makes it easy to set up a deck to mix in a new track with an intro loop.

Reply:
Well, I won't argue then. I've probably spent more time developing this mapping than actually using my controller :P

Please check if it's alright now, if yes then I need to also update the manual.

Reply:
Yes, this looks good now. Please update the manual and let's finally merge these PRs. :)